src/java/org/apache/cassandra/metrics/AccordStateCacheMetrics.java

```diff
@@ -0,0 +1,90 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.metrics;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import com.codahale.metrics.Histogram;
+import org.apache.cassandra.cache.CacheSize;
+
+import static org.apache.cassandra.metrics.CassandraMetricsRegistry.Metrics;
+
+public class AccordStateCacheMetrics extends CacheMetrics
+{
+    public final Histogram objectSize;
+
+    /**
+     * Create metrics for given cache.
+     *
+     * @param type  Type of Cache to identify metrics.
+     * @param cache Cache to measure metrics
+     */
+    public AccordStateCacheMetrics(String type, CacheSize cache)
+    {
+        super(type, cache);
+
+        objectSize = Metrics.histogram(factory.createMetricName("ObjectSize"), false);
+    }
+
+    public static class Provider
+    {
+        public static final String ACCORD_STATE_CACHE = "accord-state-cache";
+
+        private final Set<CacheSize> globalCacheSizes;
+        private final AccordStateCacheMetrics globalMetrics;
+
+        private final Map<Class<?>, AccordStateCacheMetrics> instanceMetrics;
+        private final Map<Class<?>, Set<CacheSize>> instanceCacheSizes;
+
+        private final String type;
+
+        public Provider()
+        {
+            this(ACCORD_STATE_CACHE);
+        }
+
+        @VisibleForTesting
+        public Provider(String type)
+        {
+            this.type = type;
+            globalCacheSizes = new CopyOnWriteArraySet<>();
+            globalMetrics = new AccordStateCacheMetrics(type, new CacheSize.Summing(globalCacheSizes));
+            instanceMetrics = new HashMap<>(2);
+            instanceCacheSizes = new HashMap<>(2);
+        }
+
+        public synchronized AccordStateCacheMetrics forInstance(Class<?> klass, CacheSize commandStoreInstanceCacheSize)
+        {
+            Set<CacheSize> cacheSizes = instanceCacheSizes.computeIfAbsent(klass, k -> new CopyOnWriteArraySet<>());
+            cacheSizes.add(commandStoreInstanceCacheSize);
+            return instanceMetrics.computeIfAbsent(klass, k -> new AccordStateCacheMetrics(String.format("%s-%s", type, k.getSimpleName()), new CacheSize.Summing(cacheSizes)));
```

Comment:
Looking at how the caches work at the instance level, there may be a problem w/ the summing approach here.

```
@Override
public long capacity()
{
    return AccordStateCache.this.capacity();
}

@Override
public void setCapacity(long capacity)
{
    AccordStateCache.this.setCapacity(capacity);
}

@Override
public int size()
{
    return AccordStateCache.this.size();
}

@Override
public long weightedSize()
{
    return bytesCached;
}
```
If we take the sum of all the `Instance` `size()` methods, won't we be taking the same shared size from `cache` in the top level `AccordStateCache`? (i.e. We're still actually looking at a global size?)

At this point, I'm not sure we should have `Instance` implement `CacheSize` at all, and I'm not sure having instance-level metrics is worth the complexity. The `Instance` caches share the same physical cache, and are less caches than they are cache accessors...

Reply:
Or perhaps here's another approach...

One of our problems here is that the instances caches are just views on an actual physical cache. `CacheMetrics` is really only appropriate for the global case. We could split out the bits of `CacheMetrics` (hits, misses, etc.) that are totally unambiguous for the instance "caches" and avoiding having to worry about any summing logic around those.

Reply:
Indeed, it is broken. I've made a mistake because `size()` for instance cache should return `itemsCached` rather than refer to the parent. It's a shame that my tests do not cover that.

Reply:
It actually works in a way that for the `AccordCommandStore` global metrics, the provider always returns the same metrics object; a similar approach was taken for the instance cache - for the same instance type, the same metrics are returned. We do not have to worry about summing up cache access metrics (hits, misses, requests) across different command stores because they all are reported to the same metrics instances.

For `CacheSize`, I believe we need to do summing on demand because the cache size can be individually set on each `AccordCommandStore`. Also, `size` and `weightedSize` are computed internally and separately for each command store and instance. So, whenever the operator requests the current number of items or the current number of bytes caches, for, say, instance X or global, we need to go through all the command stores (implementing `CacheSize`) and sum up the outputs.

Reply:
> We do not have to worry about summing up cache access metrics (hits, misses, requests) across different command stores because they all are reported to the same metrics instances.

Right. Understand that part.

> For CacheSize, I believe we need to do summing on demand because the cache size can be individually set on each AccordCommandStore

Again, agreed. I think where we still have things to resolve are:

1.) What's the most straightforward way to expose the summed sizes (see this thread https://github.com/apache/cassandra/pull/2534/files#r1317595176)

2.) Whether we really want to worry about exposing the `CacheSize` bits for the instance views at all. Again, we're talking about the same physical cache for all instances, right? Even if we don't do that right now in this patch, we could always add it later if there was a really compelling reason. This is why I suggested something like a `CacheAccessMetrics` that could just worry about exposing the hist/misses/etc. items for the `Instance` caches, which don't have to be summed at all.

Reply:
> they all are reported to the same metrics instances.

Should we be? We can manage much cheaper metrics per-command store. Whether we do that immediately or not, we should probably consider structuring our API to support this.

Reply:
So do you mean individual metrics per command store instead of aggregated?

Reply:
Whether we actually instantiate separate metrics per command store or not (we probably don't have optimally-efficient ones today for that in C*), it would be nice to structure the API to support the obviously more efficient approach, and move Cassandra to such an approach when ready.

Reply:
So by that I mean probably provide the CommandStore its own metrics callback on construction; but in our C* implementation this might be the same callback.

Reply:
The problem with `CacheMetrics` is that they include both "push" and "pull" items. So perhaps it is a good idea to separate the concepts of `CacheSize` and `CacheAccessMetrics`. Will think about that over the weekend;

Reply:
Ok, so I refactored things a bit (tests do not compile yet, I need to get your opinion first). I separated cache size metrics and cache access metrics and now it looks simpler to me. `AccordCommandStores` and `AccordCommandStore` both implement `CacheSize`. `AccordCommandStores` include cache size metrics and cache access metrics. The instance of the later is passed to each new `AccordCommandStore`. So in short, we have cache size metrics only for global `AccordCommandStores` while cache access metrics are both global and per instance. I hope this is sufficient

Reply:
The new changes look reasonable to me...simpler and clearer.